contracts/hooks/wormhole/src/lib.rs

```diff
@@ -0,0 +1,157 @@
+    WORMHOLE_CORE.save(deps.storage, &wormhole_core)?;
+    MAILBOX.save(deps.storage, &mailbox)?;
+
+    Ok(Response::new().add_event(
+        new_event("initialize")
+            .add_attribute("sender", info.sender)
+            .add_attribute("owner", owner)
+            .add_attribute("wormhole_core", wormhole_core),
+    ))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        // TODO: maybe add SetWormholeCore Msg
+        ExecuteMsg::Ownable(msg) => Ok(hpl_ownable::handle(deps, env, info, msg)?),
+        ExecuteMsg::PostDispatch(msg) => post_dispatch(deps, info, msg),
+    }
+}
+
+fn post_dispatch(
+    deps: DepsMut,
+    _info: MessageInfo,
+    req: PostDispatchMsg,
+) -> Result<Response, ContractError> {
+    // Ensure message_id matches latest dispatch from mailbox
+    let mailbox = MAILBOX.load(deps.storage)?;
+    let latest_dispatch_id = deps
+        .querier
+        .query_wasm_smart::<LatestDispatchedIdResponse>(
+            &mailbox,
+            &MailboxQueryMsg::LatestDispatchId {}.wrap(),
+        )?
+        .message_id;
+
+    let decoded_msg: Message = req.message.clone().into();
+
+    ensure_eq!(
+        latest_dispatch_id,
+        decoded_msg.id(),
+        ContractError::Unauthorized {}
+    );
+
+    // send message to wormhole core-bridging-contract
+    let wormhole_core = WORMHOLE_CORE.load(deps.storage)?;
+    let decoded_msg: Message = req.message.clone().into();
+    let binary_message = Binary::from(req.message); // why req.message?
+    let wormhole_message: WormholeExecuteMsg = WormholeExecuteMsg::PostMessage {
+        nonce: decoded_msg.nonce,
+        message: binary_message,
+    };
+
+    let wormhole_msg = wasm_execute(&wormhole_core, &wormhole_message, vec![])?;
+
+    Ok(Response::new().add_message(wormhole_msg).add_event(
+        new_event("post_dispatch").add_attribute("message_id", decoded_msg.id().to_hex()),
+    ))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> Result<QueryResponse, ContractError> {
+    match msg {
+        QueryMsg::Wormhole(msg) => Ok(handle_query(deps, env, msg)?),
+        QueryMsg::Ownable(msg) => Ok(hpl_ownable::handle_query(deps, env, msg)?),
+        QueryMsg::Hook(msg) => match msg {
+            HookQueryMsg::Mailbox {} => to_binary(get_mailbox(deps)),
+            HookQueryMsg::QuoteDispatch(_) => to_binary(quote_dispatch()),
+        },
+    }
+}
+
+pub fn handle_query(deps: Deps, _env: Env, _msg: WormholeQueryMsg) -> StdResult<QueryResponse> {
+    cosmwasm_std::to_json_binary(&WormholeInfoResponse {
+        wormhole_core: WORMHOLE_CORE.load(deps.storage)?.into_string(),
+    })
+}
+
+fn get_mailbox(_deps: Deps) -> Result<MailboxResponse, ContractError> {
+    Ok(MailboxResponse {
+        mailbox: "unrestricted".to_string(),
```

Comment:
Is there a point in returning "unrestricted"? Since we have MAILBOX  in state and can return actual address